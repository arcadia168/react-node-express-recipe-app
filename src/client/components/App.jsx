import React, { Component } from 'react';
import { Nav, NavbarToggler, NavLink, Navbar, NavbarBrand, NavItem, Collapse, Button } from 'reactstrap';
import '../scss/application.scss';
import Home from './Home/Home.jsx'
import history from '../services/HistoryService'

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      recipes: [],
      isOpen: false
    };
    this.toggle = this.toggle.bind(this);
  }

  toggle() {
    this.setState({
      isOpen: !this.state.isOpen
    });
  }

  goTo(route) {
    this.props.history.replace(`/${route}`)
  }

  login() {
    this.props.auth.login();
  }

  logout() {
    this.props.auth.logout();
    this.goTo('home');
  }

  render() {
    const { isAuthenticated } = this.props.auth;
    const recipeList = this.state.recipes;

    return (
      <div>
        <Navbar color="faded" light expand="xs">
          <NavbarToggler onClick={this.toggle} />
          <Collapse isOpen={this.state.isOpen} navbar>
            <Nav className="ml-auto" navbar>
              <NavItem>
                <NavLink href="/home">Home</NavLink>
              </NavItem>
              {
                !isAuthenticated() && (
                  <NavItem>
                    <NavLink onClick={this.login.bind(this)}>
                      Log In
                    </NavLink>
                  </NavItem>
                )
              }
              {
                isAuthenticated() && (
                  <NavItem>
                    <NavLink onClick={this.logout.bind(this)}>
                      Log Out
                    </NavLink>
                  </NavItem>
                )
              }
            </Nav>
          </Collapse>
        </Navbar>
      </div >);
  }
}

module.exports = App;