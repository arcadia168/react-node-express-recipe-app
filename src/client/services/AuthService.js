import auth0 from 'auth0-js';
import history from './HistoryService'
import axios from 'axios';

class AuthService {
    constructor() {
        this.auth0 = new auth0.WebAuth({
            domain: 'bennawazcodedemos.eu.auth0.com',
            clientID: 'B5bvppKVgfaM8cj3eukBOhoShRq0eBBE',
            redirectUri: 'http://localhost:3000/callback',
            audience: 'react-node-recipes-auth-api',
            responseType: 'token id_token',
            scope: 'openid profile email'
        });

        this.userProfile = undefined;
        this.login = this.login.bind(this);
        this.logout = this.logout.bind(this);
        this.handleAuthentication = this.handleAuthentication.bind(this);
        this.isAuthenticated = this.isAuthenticated.bind(this);
    }

    handleAuthentication() {
        this.auth0.parseHash((err, authResult) => {
            if (authResult && authResult.accessToken && authResult.idToken) {
                this.setSession(authResult);
                history.replace('/home');
            } else if (err) {
                history.replace('/home');
                console.log(err);
            }
        });
    }

    getUserProfile() {
        if (!this.userProfile) {
            var accessToken = localStorage.getItem('access_token');

            if (!accessToken) {
                console.log('Access token must exist to fetch profile');
            }

            this.auth0.client.userInfo(accessToken, (err, profile) => {
                if (profile) {
                    this.userProfile = profile;
                }
            });
        } else {
            return this.userProfile;
        }
    }

    updateUserProfile(user_profile) {
        axios.post('/api/users', user_profile).then((updatedUserProfile) => {
            //use the favourite recipes
            console.log('User profile updated on backend API: ' + JSON.stringify(updatedUserProfile));
            localStorage.setItem('user_profile', updatedUserProfile);
        }).catch((error) => {
            //log error
            console.log(error);
        });
    }

    setSession(authResult) {
        // Set the time that the access token will expire at
        let expiresAt = JSON.stringify((authResult.expiresIn * 1000) + new Date().getTime());
        localStorage.setItem('access_token', authResult.accessToken);
        localStorage.setItem('id_token', authResult.idToken);
        localStorage.setItem('expires_at', expiresAt);
        localStorage.setItem('user_profile', authResult.idTokenPayload);
        //set the axios service to use this auth header
        ;
        axios.defaults.headers.common['Authorization'] = `Bearer ${localStorage.getItem('access_token')}`;
        //send the user profile to the backend and store/update it in the db
        this.updateUserProfile(authResult.idTokenPayload);
        // navigate to the home route
        history.replace('/home');
    }

    logout() {
        // Clear access token and ID token from local storage
        localStorage.removeItem('access_token');
        localStorage.removeItem('id_token');
        localStorage.removeItem('expires_at');
        // navigate to the home route
        history.replace('/home');
    }

    isAuthenticated() {
        // Check whether the current time is past the 
        // access token's expiry time
        let expiresAt = JSON.parse(localStorage.getItem('expires_at'));
        return new Date().getTime() < expiresAt;
    }

    login() {
        this.auth0.authorize();
    }
}

module.exports = AuthService;