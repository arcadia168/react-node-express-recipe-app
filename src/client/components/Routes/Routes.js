import React from 'react';
import { Router, Route } from 'react-router-dom';
import App from '../App.jsx';
import Callback from '../Callback/Callback.jsx';
import HistoryService from '../../services/HistoryService';
import AuthService from '../../services/AuthService';
import axios from 'axios';

const auth = new AuthService();
const axiosInstance = axios;
axiosInstance.defaults.headers.common['Authorization'] = `Bearer ${localStorage.getItem('access_token')}`;

const handleAuthentication = (nextState, replace) => {
  if (/access_token|id_token|error/.test(nextState.location.hash)) {
    auth.handleAuthentication();
  }
}

export const makeMainRoutes = () => {
  return (
    <Router history={HistoryService} component={App}>
      <div>
        <Route path="/" render={(props) => <App axios={axiosInstance} auth={auth} {...props}/>}/>
        <Route path="/callback" render={(props) => {
          handleAuthentication(props);

          return <Callback {...props} /> 
        }}/>
      </div>
    </Router>
  );
}