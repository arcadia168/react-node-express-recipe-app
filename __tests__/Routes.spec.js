import React from 'react';
import ReactDOM from 'react-dom';
import App from 'App';
import HistoryService from 'HistoryService';
import AuthService from 'AuthService';
import axios from 'axios';
import renderer from 'react-test-renderer';
const axiosInstance = axios;
const authInstance = new AuthService(axiosInstance);
import { makeMainRoutes } from 'Routes';

describe('Routes', () => {
    it('renders correctly', () => {
        const routes = makeMainRoutes();
        const tree = renderer
          .create(routes)
          .toJSON();
        expect(tree).toMatchSnapshot();
      });
});