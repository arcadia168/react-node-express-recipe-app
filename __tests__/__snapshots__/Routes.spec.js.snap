// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routes renders correctly 1`] = `<div />`;
